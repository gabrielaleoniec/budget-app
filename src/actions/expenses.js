import database from '../firebase/firebase';

// ADD EXPENSE
export const addExpense = (expense = {}) => {
    return {
        type: "ADD_EXPENSE",
        expense
    }
}

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const date = createdAt ? new Date(createdAt).valueOf() : new Date().valueOf();

        const expense = { description, note, amount, createdAt: date };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                ...expense,
                id: ref.key
            }))
        });
    }
}

// REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id: id
});

// START REMOVE EXPENSES
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => { 
            dispatch(removeExpense({ id }))
        });
    }
}

// EDIT EXPENSE
export const updateExpense = (id, updates) => ({
    type: "UPDATE_EXPENSE",
    id,
    updates
})

// SET EXPENSES
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then(snapshot => {
            const expenses = [];

            snapshot.forEach(childSnapshot => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        }).catch((e) => {
            console.log('Error', e);
        })
    }
};
