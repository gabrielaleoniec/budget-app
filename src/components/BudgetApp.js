import React from 'react';

const BudgetApp = () => {
    return <div>Inside the compppponent</div>
}

export default BudgetApp;