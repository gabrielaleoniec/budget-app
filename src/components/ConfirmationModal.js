import React from 'react';

const ConfirmationModal = (props) => {
    const className = props.showConfirmationBox ? "confirmation-box--show" : '';
    return <div className="box-layout box-layout--cover">
        <div className={`confirmation-box ${className}`}>
            <p>Should the expense be removed?</p>
            <div className="confirmation-box__buttons">
                <button className="button button--weak" onClick={props.hideConfirmationModal}>Cancel</button>
                <button className="button button--remove" onClick={props.removeExpense}>Remove</button>
            </div>
        </div>
    </div>
}

export default ConfirmationModal;
