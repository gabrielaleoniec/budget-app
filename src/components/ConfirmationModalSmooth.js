import React from 'react';

const ConfirmationModalSmooth = () => {
    return () => {
        
    }
}