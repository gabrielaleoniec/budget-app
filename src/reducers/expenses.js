const defaultStateExpensifyReducer = [];

const expensifyReducer = (state = defaultStateExpensifyReducer, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'UPDATE_EXPENSE':
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                return expense;
            })
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state;
    }
}

export default expensifyReducer;