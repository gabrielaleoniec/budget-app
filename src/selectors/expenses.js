
const getVisibleExpenses = (expenses, { text, sortBy, sortByAsc, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const matchText = expense.description.toLowerCase().includes(text.toLowerCase());
        const matchStartDate = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const matchEndDate = typeof endDate !== 'number' || expense.createdAt <= endDate;

        return matchText && matchStartDate && matchEndDate;
    }).sort((a, b) => {
        if (sortBy === 'createdAt') {
            return sortByAsc ? b.createdAt - a.createdAt : a.createdAt - b.createdAt;
        } else if (sortBy === 'amount') {
            return sortByAsc ? b.amount - a.amount : a.amount - b.amount;
        }
    })
}

export default getVisibleExpenses;