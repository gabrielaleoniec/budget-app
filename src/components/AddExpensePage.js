import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

const AddExpensePage = (props) => (
    <div>
        <h2>Add expense</h2>
        <ExpenseForm onSubmit={(expense) => {
            props.dispatch(addExpense(expense));
            props.history.push('/');
        }}/>
    </div>
);

const mapStateToProps = (state) => ({
    expenses: state.expenses
})

export default connect(mapStateToProps)(AddExpensePage);