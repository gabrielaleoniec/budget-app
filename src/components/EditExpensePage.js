import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import ExpenseForm from './ExpenseForm';
import { startUpdateExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        const createdAt = typeof expense.createdAt === 'string' ? moment(expense.createdAt).valueOf() : expense.createdAt;

        this.props.startUpdateExpense(this.props.expense.id, { ...expense, createdAt });
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    render() {
        if (!this.props.expense || !this.props.expense.id) {
            return <p>Expense for the given id does not exist</p>
        }

        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h2 className="page-header__title">Edit expense</h2>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense} />
                    <button onClick={this.onRemove} className="button button--remove">Remove expense</button>
                </div>
            </div>
        )
    }
}
const mapDispatchToProps = (dispatch, props) => ({
    startUpdateExpense: (id, expense) => dispatch(startUpdateExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);