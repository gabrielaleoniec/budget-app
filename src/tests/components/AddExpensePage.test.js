import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { AddExpensePage } from '../../components/AddExpensePage';

let startAddExpense, history, wrapper;
beforeEach(() => {
    startAddExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage startAddExpense={startAddExpense} history={history} />);
});

test('Should render AddExpense correcty ', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should call startAddExpense prop on valid form submission', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startAddExpense).toHaveBeenLastCalledWith(expenses[0]);
});