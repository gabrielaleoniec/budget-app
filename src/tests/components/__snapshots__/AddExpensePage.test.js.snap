// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddExpense correcty  1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Add expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
